pandas/core/reshape/merge.py

```diff
@@ -1446,10 +1446,18 @@ def _get_merge_keys(self):
         # validate index types are the same
         for i, (lk, rk) in enumerate(zip(left_join_keys, right_join_keys)):
             if not is_dtype_equal(lk.dtype, rk.dtype):
-                raise MergeError("incompatible merge keys [{i}] {lkdtype} and "
-                                 "{rkdtype}, must be the same type"
-                                 .format(i=i, lkdtype=lk.dtype,
-                                         rkdtype=rk.dtype))
+                if (is_categorical_dtype(lk.dtype) and
```

Comment:
so instead of this, i would instead just use ``repr(lk.dtype)`` and same for the right; this is a slight change for all dtypes, see if any tests break.

Reply:
What do you mean exactly? Using `repr(lk.dtype) == repr(rk.dtype)` instead of the `is_dtype_equal(...)` comparison? Or instead of the `is_categorical_dtype(...)`? Or both?

Reply:
neither use repr(lk.dtype) in the formatter

Reply:
Makes more sense. Done everywhere in `_get_merge_keys()`. Adjusted the tests that matched on the MergeError exception message to ignore the exact representation of the dtype.